lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4153,87 +3965,1112 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+  }
+
+  //--------------------------------------------------------------------------//
+  // Tangent emission helpers
+  //--------------------------------------------------------------------------//
+
+  void emitTangentForDestroyValueInst(DestroyValueInst *dvi) {
+    auto &diffBuilder = getDifferentialBuilder();
+    auto loc = dvi->getLoc();
+    auto tanVal = materializeTangent(getTangentValue(dvi->getOperand()), loc);
+    diffBuilder.emitDestroyValue(loc, tanVal);
+  }
+
+  void emitTangentForBeginBorrow(BeginBorrowInst *bbi) {
+    auto &diffBuilder = getDifferentialBuilder();
+    auto loc = bbi->getLoc();
+    auto tanVal = materializeTangent(getTangentValue(bbi->getOperand()), loc);
+    auto tanValBorrow = diffBuilder.emitBeginBorrowOperation(loc, tanVal);
+    setTangentValue(bbi->getParent(), bbi,
+                    makeConcreteTangentValue(tanValBorrow));
+  }
+
+  void emitTangentForEndBorrow(EndBorrowInst *ebi) {
+    auto &diffBuilder = getDifferentialBuilder();
+    auto loc = ebi->getLoc();
+    auto tanVal = materializeTangent(getTangentValue(ebi->getOperand()), loc);
+    diffBuilder.emitEndBorrowOperation(loc, tanVal);
+  }
+
+  void emitTangentForCopyValueInst(CopyValueInst *cvi) {
+    auto &diffBuilder = getDifferentialBuilder();
+    auto tan = getTangentValue(cvi->getOperand());
+    auto tanVal = materializeTangent(tan, cvi->getLoc());
+    auto tanValCopy = diffBuilder.emitCopyValueOperation(cvi->getLoc(), tanVal);
+    setTangentValue(cvi->getParent(), cvi,
+                    makeConcreteTangentValue(tanValCopy));
+  }
+
+  void emitTangentForReturnInst(ReturnInst *ri) {
+    auto loc = ri->getOperand().getLoc();
+    auto diffBuilder = getDifferentialBuilder();
+    // This vector will contain all the materialized return elements.
+    SmallVector<SILValue, 8> retElts;
+    // This vector will contain all indirect parameter tangent buffers.
+    // TODO: Handle indirect results.
+    auto tanParam =
+        materializeTangent(getTangentValue(ri->getOperand()), loc);
+    diffBuilder.createReturn(ri->getLoc(), tanParam);
+  }
+
+  void emitTangentForApplyInst(ApplyInst *ai, SILAutoDiffIndices indices) {
+    auto *bb = ai->getParent();
+    auto loc = ai->getLoc();
+    auto diffBuilder = getDifferentialBuilder();
+
+    // Get the differential.
+    auto *field = differentialInfo.lookUpLinearMapDecl(ai);
+    assert(field);
+    SILValue differential = getDifferentialStructElement(bb, field);
+
+    SmallVector<SILValue, 8> diffArgs;
+    for (auto origArg : ai->getArguments()) {
+      // Get the tangent value of the original parameter.
+      if (!activityInfo.isActive(origArg, getIndices()))
+        continue;
+      SILValue tanParam;
+      if (origArg->getType().isObject()) {
+        // If original result is a `tuple_extract`, materialize tangent value of
+        // `ai` and extract the corresponding element tangent value.
+        if (auto *tupleExtract =
```

Comment:
No need for this `if` logic. This is dead code that has been removed from pullback emitter. You only need what's in `else`: `tanParam = materializeTangent(getTangentValue(origArg), loc);`.